File: src/parser/list-items.ts
import type { BundledTheme } from "shiki";
import parser from "./index.ts";

export function orderedListItems(input: string, codeTheme: BundledTheme) {
  const lines = input
    .split(/^\d+\. /gm)
    .filter((l) => l)
    .map((l) => l.trim());

  const tokens: HastElement[] = lines.map((line) => {
    const segments = line.split("\n");
    if (segments.length === 1) {
      return {
        type: "element",
        tagName: "li",
        children: parser(line, codeTheme),
      };
    }

    return {
      type: "element",
      tagName: "li",
      children: parser(
        `${segments[0]}\n\n${segments
          .slice(1)
          .map((l) => l.replace(/^(   |\t)/m, ""))
          .join("\n")}`,
        codeTheme,
      ),
    };
  });

  return tokens;
}

export function unorderedListItems(input: string, codeTheme: BundledTheme) {
  const lines = input
    .split(/^- /gm)
    .filter((l) => l)
    .map((l) => l.trim());
  const tokens: HastElement[] = lines.map((line) => {
    const segments = line.split("\n");
    if (segments.length === 1) {
      return {
        type: "element",
        tagName: "li",
        children: parser(line, codeTheme),
      };
    }

    return {
      type: "element",
      tagName: "li",
      children: parser(
        `${segments[0]}\n\n${segments
          .slice(1)
          .map((l) => l.replace(/^(  |\t)/m, ""))
          .join("\n")}`,
        codeTheme,
      ),
    };
  });

  return tokens;
}

File: src/parser/list_items.rs
use highlight::Theme;
use regex::Regex;

use crate::types::Node;

use super::parser::parse;

pub enum ListType {
    Ordered,
    Unordered,
}

pub fn list_items(input: String, code_theme: Theme, list_type: ListType) -> Vec<Node> {
    let (re, seg_re) = match list_type {
        ListType::Ordered => (r"(?m)^\d+\. ", r"^(   |\t)"),
        ListType::Unordered => (r"(?m)^- ", r"^(  |\t)"),
    };

    let re = match Regex::new(re) {
        Ok(re) => re,
        Err(_) => unreachable!(),
    };

    let lines = re.split(&input).filter(|l| !l.is_empty()).map(|l| l.trim());
    let tokens = lines
        .map(|line| {
            let segments = line.lines();
            if segments.clone().count() == 1 {
                return Node::element("li", None, parse(line.to_string(), code_theme.clone()));
            }

            let segments = segments.collect::<Vec<&str>>();
            return Node::element(
                "li",
                None,
                parse(
                    format!(
                        "{}\n\n{}",
                        segments[0],
                        segments[1..]
                            .into_iter()
                            .map(|l| {
                                let re = match Regex::new(seg_re) {
                                    Ok(re) => re,
                                    Err(_) => unreachable!(),
                                };
                                re.replace(l, "").to_string()
                            })
                            .collect::<Vec<String>>()
                            .join("\n")
                    ),
                    code_theme.clone(),
                ),
            );
        })
        .collect();
    tokens
}
